Add support for Mongo 6.0

mserranom opened the pull request:
## Description


Adds support for Mongo 6.0 in preparation for the next minor release. The main changes are:

- `mongo` shell has been replaced by `mongosh`. The toolkit now reads Mongo major version from `overleaf.rc` and sets a `MONGOSH` variable that is used in different places to invoke the Mongo shell.
- Replica set initialisation with `rs.initiate()` fails in Mongo 6.0 when the replica set is already initialised. It's now checking `db.isMaster().primary` beforehand, which should return an empty string if the replicaSet is not initialised.

Upgrade instructions have been updated: https://github.com/overleaf/overleaf/wiki/Updating-Mongo-version/_compare/4e81c4348e2f8a30e2676af13ca7271493f53e75...acc881bbfc8f257d8d1b016e0ecbdd6cbc12b0ce

**Test**

With `MONGO_IMAGE=mongo:5.0` in `overleaf.rc`:

```
# Image inspection
"Image": "mongo:5.0",
 "Healthcheck": {
            "Test": [
                "CMD-SHELL",
                "echo 'db.stats().ok' | mongo localhost:27017/test --quiet"
            ],
            "Interval": 10000000000,
            "Timeout": 10000000000,
            "Retries": 5
},
```

```
$ bin/mongo
MongoDB shell version v5.0.18
connecting to: mongodb://127.0.0.1:27017/sharelatex?compressors=disabled&gssapiServiceName=mongodb
```

Updated `MONGO_IMAGE=mongo:6.0` in `overleaf.rc` (previously set feature compatibility to `5.0`):

```
# Image inspection
 "Image": "mongo:6.0",
 "Healthcheck": {
            "Test": [
                "CMD-SHELL",
                "echo 'db.stats().ok' | mongosh localhost:27017/test --quiet"
            ],
            "Interval": 10000000000,
            "Timeout": 10000000000,
            "Retries": 5
        },
```

```
$ bin/mongo
Current Mongosh Log ID: 667032abd41272b4ef8db5fa
Connecting to:          mongodb://127.0.0.1:27017/sharelatex?directConnection=true&serverSelectionTimeoutMS=2000&appName=mongosh+2.2.6
Using MongoDB:          6.0.15
Using Mongosh:          2.2.6
```


In both versions: tested login, project creation and compiles.

## Contributor Agreement

- [x] I confirm I have signed the [Contributor License Agreement](https://github.com/overleaf/overleaf/blob/master/CONTRIBUTING.md#contributor-license-agreement)

das7pad commented on lib/shared-functions.sh:
```diff
@@ -19,6 +19,22 @@ function read_image_version() {
+function read_mongo_version() {
+  local mongo_image
+  mongo_image=$(read_configuration "MONGO_IMAGE")
+  if [[ ! "$mongo_image" =~ ^mongo:([0-9]+)\.(.*)$ ]]; then
```
I'm aware of customers with custom docker registries. Their images would not match this regex. WDYT about removing the name and just matching the tag?

As an alternative, we could split the `MONGO_IMAGE` variable into two parts, the name and the version.

mserranom replied:
I added a check for `MONGO_IMAGE`:

- If `MONGO_VERSION` is provided it reads the major version from there.
- If not provided, checks whether `MONGO_IMAGE` is a stock version and parses the major version.
- If `MONGO_IMAGE` is not a stock image, issues a warning and defaults to `5` for backwards compatibility.

The warning might be printed more than once (tried to add a check for `MONGOSH` existence check, but the variables are unbounded in subshells), which I think we can live with.

das7pad commented on lib/shared-functions.sh:
```diff
@@ -19,6 +19,46 @@ function read_image_version() {
 
+function read_mongo_version() {
+
+  # Reads `MONGO_VERSION` from config/overleaf.rd if to set the correct mongo shell
```
```suggestion
  # Reads `MONGO_VERSION` from config/overleaf.rc if to set the correct mongo shell
```

das7pad commented on lib/shared-functions.sh:
```diff
@@ -19,6 +19,46 @@ function read_image_version() {
+      echo "-------------------------------------------------------"
+      echo "---------------------  WARNING  -----------------------"
+      echo "-------------------------------------------------------"
+      MONGO_VERSION_MAJOR=5 # default to version 5 for backwards compatibility
```
I would prefer failing here rather than producing a cryptic error down the line.

das7pad reviewed (commented):
LGTM, not tested. There is a risk for these two to diverge now, e.g. the following is invalid but accepted by the Toolkit:

```
MONGO_VERSION=5.0.0
MONGO_IMAGE=my.registry.example.com/mongo:6.0.0
```
WDYT about adding `MONGO_VERSION` as tag to `MONGO_IMAGE`?

Something like this:
```
if MONGO_VERSION not set {
  // try to infer version and fail when using a custom image
} else {
  if MONGO_VERSION invalid {
    exit 1
  }
  MONGO_IMAGE="$MONGO_IMAGE:$MONGO_VERSION"
}
```

mserranom commented:
> GTM, not tested. There is a risk for these two to diverge now, e.g. the following is invalid but accepted by the Toolkit:
> 
> ```
> MONGO_VERSION=5.0.0
> MONGO_IMAGE=my.registry.example.com/mongo:6.0.0
> ```
> 
> WDYT about adding `MONGO_VERSION` as tag to `MONGO_IMAGE`?

Been thinking about it:
- The tag might be custom (something like `my-mongo:v6`). We can't reliably extract the version number in those cases.
- If there's an alignment problem with `MONGO_VERSION` and `MONGO_IMAGE` they will quickly see a problem before the app is started, as `mongosh` won't work.

I lean towards leaving it as it is. I've tweaked a little bit to make it fail when `MONGO_VERSION` is not defined (as you suggested), and updated the error message to make it clear  `MONGO_IMAGE` need to be a valid version (`<major>.<minor>`).

Maybe @mlevans0 can weigh in?

mlevans0 commented on lib/shared-functions.sh:
```diff
@@ -19,6 +19,45 @@ function read_image_version() {
 
+function read_mongo_version() {
+
+  # Reads `MONGO_VERSION` from config/overleaf.rc if to set the correct mongo shell
```
Remove the 'if'

mserranom replied:
```suggestion
  # Reads `MONGO_VERSION` from config/overleaf.rc to set the correct mongo shell
```

mlevans0 commented:
If I understand correctly, the `MONGO_VERSION` environment variable has been introduced to help identify which MongoDB shell to use (either `mongo` or `mongosh`). Other than that, its not used (still using `MONGO_IMAGE` in the compose file)?

The `read_mongo_version` function attempts to read `MONGO_VERSION` from **config/overleaf.rc** to set the correct MongoDB shell, but if `MONGO_VERSION` is not defined, it then tries to extract the version from `MONGO_IMAGE` instead and then issues a warning if it can't (in cases where it's a none-stock image). 

Because of how this is handled, I think the `MONGO_VERSION` environment variable is unnecessary. Couldn't we parse the `MONGO_IMAGE` for the version number and show the warning if it's not a stock image (shortcutting the need for the variable)?

Adding `MONGO_VERSION` to the **config/overleaf.rc** file will be a manual step for existing deployments, and if missed, would result in it being undefined, and therefore falling back to the `MONGO_IMAGE` check anyway.

If we go with what's planned, and the `MONGO_VERSION` is only used for setting the MongoDB shell, this needs to be explicitly documented. It could cause some confusion when attempting upgrades - you only need to set `MONGO_IMAGE` and it must contain the version number - after MongoDB 6, you only need to set `MONGO_IMAGE` (with version number),  `MONGO_VERSION` can be anything as long as its >= 6). It's also **only** really needed when going from 5 -> 6, after that there's no future requirement for it is there?

das7pad commented:
> * If there's an alignment problem with `MONGO_VERSION` and `MONGO_IMAGE` they will quickly see a problem before the app is started, as `mongosh` won't work.

That is true for the 5-to-6 upgrade, but an upcoming `mongo:7.0` may come without breaking changes on the `mongosh`/command side. And then it will be trickier to spot that the two variables diverged.

I still like the idea of building the effective "mongo container image" using `${MONGO_IMAGE}:${MONGO_VERSION}` -- a we do with the Server Pro/git-bridge image. We could be strict about what customers can use as `MONGO_VERSION` with a custom mongo image when using the Toolkit. They can always re-tag their custom, non-standard tag.

Side note: The official `mongo` image defines a few environment variables that might be handy. But then I do not know what customers might use as custom `mongo` image -- i.e. theirs might not define these.

```
$ docker run --rm --entrypoint env mongo:6.0 | grep MONGO
MONGO_PACKAGE=mongodb-org
MONGO_REPO=repo.mongodb.org
MONGO_MAJOR=6.0
MONGO_VERSION=6.0.16

$ docker run --rm --entrypoint env mongo:5.0 | grep MONGO
MONGO_PACKAGE=mongodb-org
MONGO_REPO=repo.mongodb.org
MONGO_MAJOR=5.0
MONGO_VERSION=5.0.27

```

mserranom commented:
> That is true for the 5-to-6 upgrade, but an upcoming mongo:7.0 may come without breaking changes on the mongosh/command side. And then it will be trickier to spot that the two variables diverged.

That's a good point.

> I still like the idea of building the effective "mongo container image" using ${MONGO_IMAGE}:${MONGO_VERSION} -- a we do with the Server Pro/git-bridge image. 

I agree it's a better idea, but I mainly wanted to mitigate backwards compatibility as much as possible. I'll give it another try. The environment variables might come useful even if not defined, since in those cases we know it's not a stock image.

mserranom commented:
I've updated the PR, the new behaviour is:

1. When `MONGO_IMAGE` and `MONGO_VERSION` are both present:
  a. If `MONGO_VERSION` doesn't start with `<major>.` prints an error message indicating the correct format.
  b. Joins `"$MONGO_IMAGE:$MONGO_VERSION"` to form `$MONGO_DOCKER_IMAGE`, which is used as docker image. 

3. When `MONGO_VERSION` is not present:
  a. If it's using a stock mongo image with a version number (such as `6.0`) it prints a warning and continues the execution using `MONGO_IMAGE` as `MONGO_DOCKER_IMAGE`
  b. If it's not a stock image prints an error.

#### Manual test

With `MONGO_IMAGE=5.0`:

- [x] confirm `bin/mongo` works 

```
$ bin/docker-compose up mongo -d
-------------------  WARNING  ----------------------
  Deprecation warning: the mongo image is now split between MONGO_IMAGE
  and MONGO_VERSION configurations. Please update your config/overleaf.rc as
  your current configuration may stop working in future versions of the toolkit.
  Example: MONGO_IMAGE=mongo
           MONGO_VERSION=6.0
[+] Running 1/1
 ✔ Container mongo  Started                  
```

With `MONGO_IMAGE=6.0`:

- [x] confirm `bin/mongo` works 

```
$ bin/docker-compose up mongo -d
-------------------  WARNING  ----------------------
  Deprecation warning: the mongo image is now split between MONGO_IMAGE
  and MONGO_VERSION configurations. Please update your config/overleaf.rc as
  your current configuration may stop working in future versions of the toolkit.
  Example: MONGO_IMAGE=mongo
           MONGO_VERSION=6.0
[+] Running 1/1
 ✔ Container mongo  Started                  
```

With `MONGO_IMAGE=custom-mongo:6.0`:

```
$ bin/docker-compose up mongo -d
---------------------  ERROR  -----------------------
  Invalid MONGO_VERSION: custom

  MONGO_VERSION must start with the actual major version of mongo, followed by a dot.
  Example: MONGO_IMAGE=my.dockerhub.com/custom-mongo
           MONGO_VERSION=6.0.hotfix-1

```

With `MONGO_IMAGE=mongo` and `MONGO_VERSION=6.0`

- [x] confirm `bin/mongo` works 
- [x] no error/warning messages 


With `MONGO_IMAGE=mongo` and `MONGO_VERSION=custom`

```
$ bin/docker-compose up mongo -d
---------------------  ERROR  -----------------------
  Invalid MONGO_VERSION: custom

  MONGO_VERSION must start with the actual major version of mongo, followed by a dot.
  Example: MONGO_IMAGE=my.dockerhub.com/custom-mongo
           MONGO_VERSION=6.0.hotfix-1
```

das7pad commented on lib/shared-functions.sh:
```diff
@@ -19,6 +19,54 @@ function read_image_version() {
+      echo "  Please update your config/overleaf.rc."
+      echo ""
+      echo "  MONGO_VERSION must start with the actual major version of mongo, followed by a dot."
+      echo "  Example: MONGO_IMAGE=my.dockerhub.com/mongo"
```
```suggestion
      echo "  Example: MONGO_IMAGE=my.dockerhub.com/custom-mongo"
```
Bonus: WDYT about aligning the two examples?

mserranom replied:
Sure, done. My idea was to align with the situation causing the error, rather than with other messages, but there's no much difference and may be less confusing for devs.

das7pad commented on lib/shared-functions.sh:
```diff
@@ -19,6 +19,54 @@ function read_image_version() {
+      echo ""
+      echo "  MONGO_VERSION must start with the actual major version of mongo, followed by a dot."
+      echo "  Example: MONGO_IMAGE=my.dockerhub.com/custom-mongo"
+      echo "           MONGO_VERSION=6.0.hotfix-1"
```
```suggestion
      echo "           MONGO_VERSION=6.0-custom"
```
Bonus: WDYT about aligning the two examples?

das7pad commented on lib/shared-functions.sh:
```diff
@@ -19,6 +19,54 @@ function read_image_version() {
+        echo "  your current configuration may stop working in future versions of the toolkit."
+        echo "  Example: MONGO_IMAGE=mongo"
+        echo "           MONGO_VERSION=6.0"
+      fi
```
```suggestion
      echo "------------------- WARNING ----------------------"
      fi
```
Bonus: Adding a footer might make this easier to read (same for the error entries).

```
toolkit$  bin/docker-compose ps
-------------------  WARNING  ----------------------
  Deprecation warning: the mongo image is now split between MONGO_IMAGE
  and MONGO_VERSION configurations. Please update your config/overleaf.rc as
  your current configuration may stop working in future versions of the toolkit.
  Example: MONGO_IMAGE=mongo
           MONGO_VERSION=6.0
WARN[0000] toolkit/lib/docker-compose.base.yml: `version` is obsolete 
WARN[0000] toolkit/lib/docker-compose.vars.yml: `version` is obsolete 
...
```

mserranom replied:
Agree, this is now done.

das7pad reviewed (commented):
LGTM, thanks for going the extra mile here :tada:

It looks like there is one more edge case that needs an extra validation step to catch a tag in the image when a version is provided as well:
```
MONGO_IMAGE=my-mongo:5.0
MONGO_VERSION=5.0
```
```
toolkit$ bin/docker-compose config | grep my-mongo
    image: my-mongo:5.0:5.0
```

I've left a few bonus comments for cosmetic details.

mserranom commented:
> It looks like there is one more edge case that needs an extra validation step to catch a tag in the image when a version is provided as well:

I expected the docker error message to be enough:

```
$ bin/up mongo -d
Initiating Mongo replica set...
unable to get image 'mongo:6.0:6.0': Error response from daemon: invalid reference format
```

I guess we can add docker image validation to the script set, there's an interesting RegExp circulating in the internets: https://stackoverflow.com/questions/39671641/regex-to-parse-docker-tag 😃

das7pad commented:
> I guess we can add docker image validation to the script set, there's an interesting RegExp circulating in the internets: https://stackoverflow.com/questions/39671641/regex-to-parse-docker-tag 😃

:see_no_evil: On a glance, checking that the `MONGO_IMAGE` does not have a tag when `MONGO_VERSION` is defined should be enough. :detective:

mserranom commented:
> 🙈 On a glance, checking that the `MONGO_IMAGE` does not have a tag when `MONGO_VERSION` is defined should be enough. 🕵️

Discussed separately, this is not trivial as image names can contain URLs with ports. Going with the current solution.

das7pad reviewed (approved):
As discussed on Slack, handling ports in the image name makes the check for not having a tag in the image name non trivial. So happy to go with the checks as is.
